man/pulseaudio-dlna.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.14.
.TH PULSEAUDIO-DLNA "1" "janvier 2021" "pulseaudio-dlna 0.6.1" "User Commands"
.SH NAME
pulseaudio-dlna \- Stream audio to DLNA devices and Chromecasts
.SH DESCRIPTION
.SS "Usage:"
.TP
pulseaudio\-dlna [\-\-host <host>] [\-\-port <port>][\-\-encoder <encoders> | \fB\-\-codec\fR <codec>] [\-\-bit\-rate=<rate>]
[\-\-encoder\-backend <encoder\-backend>]
[\-\-filter\-device=<filter\-device>]
[\-\-renderer\-urls <urls>]
[\-\-request\-timeout <timeout>]
[\-\-chunk\-size <chunk\-size>]
[\-\-msearch\-port=<msearch\-port>] [\-\-ssdp\-mx <ssdp\-mx>] [\-\-ssdp\-ttl <ssdp\-ttl>] [\-\-ssdp\-amount <ssdp\-amount>]
[\-\-cover\-mode <mode>]
[\-\-auto\-reconnect]
[\-\-debug]
[\-\-fake\-http10\-content\-length] [\-\-fake\-http\-content\-length]
[\-\-disable\-switchback] [\-\-disable\-ssdp\-listener] [\-\-disable\-device\-stop] [\-\-disable\-workarounds] [\-\-disable\-mimetype\-check]
.TP
pulseaudio\-dlna [\-\-host <host>] [\-\-create\-device\-config] [\-\-update\-device\-config]
[\-\-msearch\-port=<msearch\-port>] [\-\-ssdp\-mx <ssdp\-mx>] [\-\-ssdp\-ttl <ssdp\-ttl>] [\-\-ssdp\-amount <ssdp\-amount>]
.IP
pulseaudio\-dlna [\-h | \fB\-\-help\fR | \fB\-\-version]\fR
.SH OPTIONS
.TP
\fB\-\-create\-device\-config\fR
Discovers all devices in your network and write a config for them.
That config can be editied manually to adjust various settings.
You can set:
.TP
\- Device name
\- Codec order (The first one is used if the encoder binary is available on your system)
\- Various codec settings such as the mime type, specific rules or
.TP
the bit rate (depends on the codec)
A written config is loaded by default if the \fB\-\-encoder\fR and \fB\-\-bit\-rate\fR options are not used.
.TP
\fB\-\-update\-device\-config\fR
Same as \fB\-\-create\-device\-config\fR but preserves your existing config from being overwritten
.TP
\fB\-\-host=\fR<host>
Set the server ip.
.TP
\fB\-p\fR \fB\-\-port=\fR<port>
Set the server port [default: 8080].
.TP
\fB\-e\fR \fB\-\-encoder=\fR<encoders>
Deprecated alias for \fB\-\-codec\fR
.TP
\fB\-c\fR \fB\-\-codec=\fR<codecs>
Set the audio codec.
Possible codecs are:
.TP
\- mp3
MPEG Audio Layer III (MP3)
.TP
\- ogg
Ogg Vorbis (OGG)
.TP
\- flac
Free Lossless Audio Codec (FLAC)
.TP
\- wav
Waveform Audio File Format (WAV)
.TP
\- opus
Opus Interactive Audio Codec (OPUS)
.TP
\- aac
Advanced Audio Coding (AAC)
.TP
\- l16
Linear PCM (L16)
.TP
\fB\-\-encoder\-backend=\fR<encoder\-backend>
Set the backend for all encoders.
Possible backends are:
.TP
\- generic (default)
\- ffmpeg
\- avconv
.TP
\fB\-b\fR \fB\-\-bit\-rate=\fR<rate>
Set the audio encoder's bitrate.
.TP
\fB\-\-filter\-device=\fR<filter\-device>
Set a name filter for devices which should be added.
Devices which get discovered, but won't match the
filter text will be skipped.
.TP
\fB\-\-renderer\-urls=\fR<urls>
Set the renderer urls yourself. no discovery will commence.
.TP
\fB\-\-request\-timeout=\fR<timeout>
Set the timeout for requests in seconds [default: 15].
.TP
\fB\-\-chunk\-size=\fR<chunk\-size>
Set the stream's chunk size [default: 4096].
.TP
\fB\-\-ssdp\-ttl=\fR<ssdp\-ttl>
Set the SSDP socket's TTL [default: 10].
.TP
\fB\-\-ssdp\-mx=\fR<ssdp\-mx>
Set the MX value of the SSDP discovery message [default: 3].
.TP
\fB\-\-ssdp\-amount=\fR<ssdp\-amount>
Set the amount of SSDP discovery messages being sent [default: 5].
.TP
\fB\-\-msearch\-port=\fR<msearch\-port>
Set the source port of the MSEARCH socket [default: random].
.TP
\fB\-\-cover\-mode=\fR<mode>
Set the cover mode [default: default].
Possible modes are:
.TP
\- disabled
No icon is shown
.TP
\- default
The application icon is shown
.TP
\- distribution
The icon of your distribution is shown
.TP
\- application
The audio application's icon is shown
.TP
\fB\-\-debug\fR
enables detailed debug messages.
.TP
\fB\-\-auto\-reconnect\fR
If set, the application tries to reconnect devices in case the stream collapsed
.TP
\fB\-\-fake\-http\-content\-length\fR
If set, the content\-length of HTTP requests will be set to 100 GB.
.TP
\fB\-\-disable\-switchback\fR
If set, streams won't switched back to the default sink if a device disconnects.
.TP
\fB\-\-disable\-ssdp\-listener\fR
If set, the application won't bind to the port 1900 and therefore the automatic discovery of new devices won't work.
.TP
\fB\-\-disable\-device\-stop\fR
If set, the application won't send any stop commands to renderers at all
.TP
\fB\-\-disable\-workarounds\fR
If set, the application won't apply any device workarounds
.TP
\fB\-\-disable\-mimetype\-check\fR
If set, the application won't check the device's mime type capabilities
.TP
\fB\-v\fR \fB\-\-version\fR
Show the version.
.TP
\fB\-h\fR \fB\-\-help\fR
Show the help.
.SH EXAMPLES
.IP
\- pulseaudio\-dlna
.IP
will start pulseaudio\-dlna on port 8080 and stream your PulseAudio streams encoded with mp3.
.IP
\- pulseaudio\-dlna \-\-encoder ogg
.IP
will start pulseaudio\-dlna on port 8080 and stream your PulseAudio streams encoded with Ogg Vorbis.
.IP
\- pulseaudio\-dlna \-\-port 10291 \-\-encoder flac
.IP
will start pulseaudio\-dlna on port 10291 and stream your PulseAudio streams encoded with FLAC.
.IP
\- pulseaudio\-dlna \-\-filter\-device 'Nexus 5,TV'
.IP
will just use devices named Nexus 5 or TV even when more devices got discovered.
.IP
\- pulseaudio\-dlna \-\-renderer\-urls http://192.168.1.7:7676/smp_10_
.IP
won't discover upnp devices by itself. Instead it will search for upnp renderers
at the specified locations. You can specify multiple locations via urls
separated by comma (,). Most users won't ever need this option, but since
UDP multicast packages won't work (most times) over VPN connections this is
very useful if you ever plan to stream to a UPNP device over VPN.
.SH "SEE ALSO"
The full documentation for
.B pulseaudio-dlna
is maintained as a Texinfo manual.  If the
.B info
and
.B pulseaudio-dlna
programs are properly installed at your site, the command
.IP
.B info pulseaudio-dlna
.PP
should give you access to the complete manual.


CHANGELOG.md
# Changelog

 * __master__ - (_2017-04-06_)
    - Fixed a bug where the detection of DLNA devices failed when there were multiple network interfaces
    - The application now binds to all interfaces by default
    - When using multiple network interfaces the appropriate network address is being used for streaming (new dependency `python-pyroute2` (preferred) or `python-netaddr` (fallback))
    - Migrated to GI bindings (removed dependencies `python-gobject` `python-rsvg` `python-gtk2`, new dependency `python-gi`, new optional dependencies `gir1.2-rsvg-2.0`, `gir1.2-gtk-3.0`)
    - Fixed a bug where devices with the same name could keep updating each other
    - Fixed a bug where codec bit rates could not be set although those were supported
    - Fixed a bug where a missing xml attribute prevented xml parsing
    - Added the `--disable-mimetype-check` option
    - Disabled mimetype check for virtual _Raumfeld_ devices
    - Subprocesses now always exit gracefully
    - Added the `--chunk-size` option
    - Added _pulseaudio_ as an encoder backend (*experimental*)
    - You can now just start one instance of pulseaudio-dlna
    - Fixed a bug where non-ascii characters in $PATH broke `distutils.spawn.find_executable()`
    - Also use environment's `XDG_RUNTIME_DIR` for detecting the DBus socket
    - The detection of the stream servers host address now uses the systems routing table
    - Because of devices with a kernel < 3.9, `python-zeroconf >= 0.17.4` is now required

 * __0.5.2__ - (_2016-04-01_)
    - Catched an exception when record processes cannot start properly

 * __0.5.1__ - (_2016-04-01_)
    - Fixed the `--filter-device` option
    - Prioritize _mp3_ over _flac_ for Chromecasts

 * __0.5.0.1__ - (_2016-03-09_)
    - Readded manpage

 * __0.5.0__ - (_2016-03-09_)
    - Set Yamaha devices to the appropriate mode before playing (thanks to [hlchau](https://github.com/hlchau)) (new dependency: `python-lxml`)
    - Fixed a bug where some SSDP messages could not get parsed correctly
    - Also support media renderers identifying as `urn:schemas-upnp-org:device:MediaRenderer:2`
    - Added the `--disable-workarounds` flag
    - Added the `--auto-reconnect` flag
    - Added the `--encoder-backend` option (new optional dependencies `ffmpeg`, `libav-tools`)
    - Removed shared encoder processes
    - Increased the default HTTP timeout to 15 seconds
    - Fixed a bug where manually added renderers could appear twice
    - Added device state polling for devices which start playing on their own
    - Added the flac encoder for _Google Chromecast_
    - Added support for _Google Cast Groups_ (new dependency `python-zeroconf`)
    - Removed dependency `python-beautifulsoup`
    - Fixed a bug where bytes were not decoded properly to unicode

 * __0.4.7__ - (_2015-11-18_)
    - The application can now co-exist with other applications which are using the port 1900/udp (thanks to [klaernie](https://github.com/klaernie))
    - Fixed the daemon mode to support `psutil` 1.x and 2.x (thanks to [klaernie](https://github.com/klaernie))
    - HTML entities in device descriptions are now converted automatically
    - Faster and more reliable device discovery (new dependency `python-netifaces`)
    - Added the `--cover-mode` option, one mode requires (optional) dependencies `gtk`, `cairo`, `rsvg`
    - L16 codecs are now selected better (e.g. needed for _XBox 360_)
    - Fixed a bug where sometimes it was tried to remove sinks twice on cleanup
    - Added the `--update-device-config` flag
    - Added the `--ssdp-ttl`, `--ssdp-mx`, `--ssdp-amount` options
    - Added the `--msearch-port` option

 * __0.4.6__ - (_2015-10-17_)
    - Added support for _Google Chromecast Audio_ (thanks to [leonhandreke](https://github.com/leonhandreke))
    - Fixed a bug where devices which does not specifiy control urls made the application crash
    - Added the `--disable-device-stop` flag
    - Added the `--request-timeout` option
    - You can now also add rules to renderers (e.g. `DISABLE_DEVICE_STOP`, `REQUEST_TIMEOUT`)
    - Fixed a bug where stream urls where not parsed correctly
    - Fixed a bug which made a Chomecast Audio throwing exceptions while stopping
    - Fixed a bug where the system's default encoding could not be determined when piping the applications output

 * __0.4.5.2__ - (_2015-09-21_)
    - Fixed a bug where the encoding of SSDP headers was not detected correctly (new dependency: `python-chardet`)

 * __0.4.5.1__ - (_2015-09-20_)
    - Added a missing dependency `python-concurrent.futures` (thanks to [Takkat-Nebuk](https://github.com/Takkat-Nebuk))

 * __0.4.5__ - (_2015-09-20_)
    - Exceptions while updating sink and device information from pulseaudio are now handled better
    - Changed `--fake-http10-content-length` flag to `--fake-http-content-length` to also support HTTP 1.1 requests
    - Fixed a bug where the supported device mime types could not get parsed correctly
    - Fixed a bug where the device UUID was not parsed correctly
    - Fixed a bug where just mime types beginning with `audio/` where accepted, but not e.g. `application/ogg`
    - The stream server will now respond with 206 when receiving requests with `range` header
    - UPNP control commands have now a timeout of 10 seconds
    - Fixed a bug where the wrong stream was removed from the stream manager
    - Fixed several bugs caused by purely relying on stopping actions for the devices idle state
    - Added L16 Encoder
    - The encoder option can now handle multiple options separated by comma
    - Added the `--create-device-config` flag
    - Fixed a bug where the dbus session was bound from the wrong process
    - Fix a bug where the wrong device UDN was retrieved from XML documents containing multiple devices

 * __0.4.4__ - (_2015-08-07_)
    - Added `--disable-ssdp-listener` option
    - Fixed a bug with applications which remove and re-add streams all the time
    - Added a missing dependency `python-psutil`


pulseaudio-dlna.py
#!/usr/bin/python

# This file is part of pulseaudio-dlna.

# pulseaudio-dlna is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# pulseaudio-dlna is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with pulseaudio-dlna.  If not, see <http://www.gnu.org/licenses/>.

'''
Usage:
    pulse-dlna.py --host <host> [--port <port>] [--encoder <encoder>]
    pulse-dlna.py [-h | --help | --version]

Options:
       --host=<host>        set the server ip.
    -p --port=<port>        set the server port [default: 8080].
    -e --encoder=<encoder>  set the server port [default: lame].
                            encoders are:
                              - lame  MPEG Audio Layer III (MP3)
                              - ogg   Ogg Vorbis
                              - flac  Free Lossless Audio Codec (FLAC)
                              - wav   Waveform Audio File Format (WAV)
    -v --version            show the version.
    -h --help               show the help.
    --debug                 enable debugging.
'''

import gobject
import dbus
import dbus.mainloop.glib
import multiprocessing
import logging
import sys

import docopt
import upnp.discover
import upnp.server
import pulseaudio


def main():
    logging.basicConfig(level=logging.INFO)
    options = docopt.docopt(__doc__, version='0.1')

    if not options['--host']:
        print('You must specify host address!')
        sys.exit(1)
    else:
        host = str(options['--host'])
        port = int(options['--port'])

    dlna_discover = upnp.discover.UpnpMediaRendererDiscover(host)
    dlna_discover.search()

    dlna_server = upnp.server.ThreadedDlnaServer(
        host, port, encoder=options['--encoder'])

    server_url = dlna_server.get_server_url()
    upnp_devices = []
    for upnp_device in dlna_discover.renderers:
        upnp_device.set_server_url(server_url)
        upnp_devices.append(upnp_device)
        logging.info('found upnp_device "{}"'.format(upnp_device))

    dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
    pulse = pulseaudio.PulseWatcher()
    pulse.set_upnp_devices(upnp_devices)

    dlna_server.set_bridges(pulse.bridges)
    process = multiprocessing.Process(target=dlna_server.serve_forever)
    process.start()

    try:
        mainloop = gobject.MainLoop()
        mainloop.run()
    except KeyboardInterrupt:
        print('interrupted')
    finally:
        pulse.cleanup()
        dlna_server.server_close()

if __name__ == '__main__':
    main()
